sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -325,200 +329,205 @@ export class BatchingReceiverLite {
-            // app crashes in receiveAndDelete mode.
-            if (totalWaitTimer) clearTimeout(totalWaitTimer);
-            const remainingWaitTimeInMs = getRemainingWaitTimeInMs();
-            totalWaitTimer = setTimeout(() => {
-              logger.verbose(
-                `${loggingPrefix} Batching, waited for ${remainingWaitTimeInMs} milliseconds after receiving the first message.`
-              );
-              finalAction();
-            }, remainingWaitTimeInMs);
-          }
-        }
+    const onError: OnAmqpEvent = (context: EventContext) => {
+      const eventType = context.session?.error != null ? "session_error" : "receiver_error";
+      let error = context.session?.error || context.receiver?.error;
 
-        try {
-          const data: ServiceBusMessageImpl = this._createServiceBusMessage(context);
-          if (brokeredMessages.length < args.maxMessageCount) {
-            brokeredMessages.push(data);
-          }
-        } catch (err) {
-          const errObj = err instanceof Error ? err : new Error(JSON.stringify(err));
-          logger.logError(
-            err,
-            `${loggingPrefix} Received an error while converting AmqpMessage to ServiceBusMessage`
-          );
-          reject(errObj);
-        }
-        if (brokeredMessages.length === args.maxMessageCount) {
-          finalAction();
-        }
-      };
+      if (error) {
+        error = translateServiceBusError(error);
+        logger.logError(error, `${loggingPrefix} '${eventType}' event occurred. Received an error`);
+      } else {
+        error = new ServiceBusError("An error occurred while receiving messages.", "GeneralError");
+      }
+      reject(error);
+    };
 
-      const onClose: OnAmqpEventAsPromise = async (context: EventContext) => {
-        const type = context.session?.error != null ? "session_closed" : "receiver_closed";
-        const error = context.session?.error || context.receiver?.error;
+    this._closeHandler = (error?: AmqpError | Error): void => {
+      if (
+        // no error, just closing. Go ahead and return what we have.
+        error == null ||
+        // Return the collected messages if in ReceiveAndDelete mode because otherwise they are lost forever
+        (this._receiveMode === "receiveAndDelete" && brokeredMessages.length)
+      ) {
+        logger.verbose(
+          `${loggingPrefix} Closing. Resolving with ${brokeredMessages.length} messages.`
+        );
 
-        if (error) {
-          logger.logError(error, `${loggingPrefix} '${type}' event occurred. The associated error`);
-        }
-      };
+        return resolveAfterPendingMessageCallbacks(brokeredMessages);
+      }
 
-      // Action to be performed on the "receiver_drained" event.
-      const onReceiveDrain: OnAmqpEvent = () => {
-        receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
-        receiver.drain = false;
+      reject(translateServiceBusError(error));
```

Comment:
Sorry, why is #13093 not applicable here?

Reply:
Just above this (in the `if`) we do check the receiveMode (ie, this is an example of a case where we do already do special casing).

Reply:
Oh, it's "or" in the if check. 
(I just got confused that we are always throwing an error if there is one.)